internal/daemon/controller/common/scopeids/scope_ids.go

```diff
@@ -353,3 +353,143 @@ func GetListingScopeIds(
+	resourceType resource.Type, // The type of resource being listed
+	recursive bool, // Whether the search is recursive.
+) (map[string]*scopes.ScopeInfo, error) {
+	const op = "scopeids.AuthorizedForList"
+
+	// Validation
+	switch {
+	case resourceType == resource.Unknown:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "unknown resource")
+	case iamRepoFn == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "nil iam repo")
+	case rootScopeId == "":
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing root scope id")
+	case authResults.Scope == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "nil scope in auth results")
+	case !recursive && authResults.Scope.Id != rootScopeId:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "non-recursive search but auth results scope does not match root scope")
+	}
+
+	// This will be used to save scope info so we can put the right scope
+	// info for each returned value
+	scopeResourceMap := make(map[string]*scopes.ScopeInfo)
+
+	// Base case: if not recursive, return the scope we were given and the
+	// already-looked-up info.
+	if !recursive {
+		return map[string]*scopes.ScopeInfo{authResults.Scope.Id: authResults.Scope}, nil
+	}
+
+	repo, err := iamRepoFn()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get all scopes recursively. Start at global because we need to take into
+	// account permissions in parent scopes even if they want to scale back the
+	// returned information to a child scope and its children.
+	scps, err := repo.ListScopesRecursively(ctx, scope.Global.String())
+	if err != nil {
+		return nil, err
+	}
+
+	var deferredScopes []*iam.Scope
+	var globalHasList bool // Store whether global has list permission
+	res := perms.Resource{Type: resourceType}
+
+	// For each scope, see if we have permission to list that resource in that scope
+	for _, scp := range scps {
+		scpId := scp.GetPublicId()
+		res.ScopeId = scpId
+		aSet := authResults.FetchActionSetForType(ctx,
+			resource.Unknown, // This is overridden by `WithResource` option.
+			action.ActionSet{action.List},
+			auth.WithResource(&res),
+		)
+
+		// We only expect the action set to be nothing, or list. In case
+		// this is not the case, we bail out.
+		switch len(aSet) {
+		case 0:
+			// Defer until we've read all scopes. We do this because if the
+			// ordering coming back isn't in parent-first ordering our map
+			// lookup might fail.
+			deferredScopes = append(deferredScopes, scp)
+		case 1:
+			if aSet[0] != action.List {
+				return nil, errors.New(ctx, errors.Internal, op, "unexpected action in set")
+			}
+			if scopeResourceMap[scpId] == nil {
+				scopeResourceMap[scpId] = &scopes.ScopeInfo{
+					Id:            scp.GetPublicId(),
+					Type:          scp.GetType(),
+					Name:          scp.GetName(),
+					Description:   scp.GetDescription(),
+					ParentScopeId: scp.GetParentId(),
+				}
```

Comment:
We can come back to this, but just to make a note of it, I think ideally this transformation would happen elsewhere within the application services, and this function would instead return a `map[string]*iam.Scope`

Reply:
Sure, I'll link this comment over to the LLB when this branch is merged